src/include/daos_pipeline.h

```diff
@@ -0,0 +1,261 @@
+	 *   -- constant:
+	 *          DAOS_FILTER_CONST:	Filter object is a constant
+	 */
+	int		filter_type;
+	/**
+	 * Type of data. Only relevant for keys and constant filter type objects:
+	 *          DAOS_FILTER_TYPE_BINARY
+	 *          DAOS_FILTER_TYPE_STRING
+	 *          DAOS_FILTER_TYPE_INTEGER
+	 *          DAOS_FILTER_TYPE_REAL
+	 */
+	int		data_type;
+	/**
+	 * Number of parameters for this filter object. For example, for '=='
+	 * we have 2 parameters.
+	 */
+	uint32_t	num_params;
+	/**
+	 * If filtering by akey, this tells us which one.
+	 */
+	d_iov_t		akey;
+	/**
+	 * How many constants we have in \a constant
+	 */
+	size_t		num_constants;
+	/**
+	 * This object holds the value of the constants
+	 */
+	d_iov_t		*constant;
+	/**
+	 * If filter should only be applied starting at an offset of the data.
+	 */
+	size_t		data_offset;
+	/**
+	 * Size of the data to be filtered.
+	 */
+	size_t		data_len;
+} daos_pipeline_filter_t;
+
+/**
+ * A pipeline node, used to build a pipeline.
+ */
+typedef struct {
+	/**
+	 * Node type can be any of the following:
+	 *   -- DAOS_PIPELINE_CONDITION:
+	 *          Records in, and records (meeting condition) out
+	 *   -- DAOS_PIPELINE_AGGREGATION:
+	 *          Records in, a single value out
+	 *
+	 * NOTE: Pipeline nodes can only be chained the following way:
+	 *             (condition) --> (condition)
+	 *             (condition) --> (aggregation)
+	 */
+	int			node_type;
+	/**
+	 * Number of filters inside this pipeline node
+	 */
+	size_t			num_filters;
+	/**
+	 * Array of filters for this node
+	 */
+	daos_pipeline_filter_t	*filters;
+} daos_pipeline_node_t;
+
+/**
+ * A pipeline.
+ */
+typedef struct {
+	/**
+	 * Version number of the data structure.
+	 */
+	uint64_t		version;
+	/**
+	 * Number of nodes chained in this pipeline
+	 */
+	size_t			num_nodes;
+	/**
+	 * Array of nodes for this pipeline
+	 */
+	daos_pipeline_node_t	*nodes;
+} daos_pipeline_t;
+
+
+/**
+ * Adds a new pipeline node to the pipeline 'pipeline' object. The effect of
+ * this function is to "push back" the new node at the end of the pipeline.
+ *
+ * \param[in,out]	pipeline	Pipeline object.
+ *
+ * \param[in]		node		Node object to be added to the pipeline.
+*/
+int
+daos_pipeline_push(daos_pipeline_t *pipeline, daos_pipeline_node_t *node);
+
+/**
+ * Adds a new filter object to the pipeline node object 'node'.
+ *
+ * \param[in,out]	node	Pipeline node object.
+ *
+ * \param[in]		filter	Filter object to be added to a pipeline node.
+ */
+int
+daos_pipeline_node_push(daos_pipeline_node_t *node,
+			daos_pipeline_filter_t *filter);
+
+/**
+ * Checks that a pipeline object is well built. If the pipeline object is well
+ * built, the function will return 0 (no error).
+ *
+ * \param[in]		pipeline	Pipeline object.
+ */
+int
+daos_pipeline_check(daos_pipeline_t *pipeline);
+
+/**
+ * Runs a pipeline on DAOS, returning objects and/or aggregated results.
+ *
+ * \param[in]		oh		Object open handle.
```

Comment:
We will never want to iterate over the object ID too? Even if we don't need it for now, i think that it would be great to reserve something in the API for this capability since we might need it for things like find in the future ...

Reply:
@johannlombardi OK, I will think about this.

Reply:
We will add the container ID to the API, and make the object ID optional. When the object ID is optional, we can filter by Object ID (I will create a new filter type to do just that).